testing/xfpga/test_metrics_c.cpp

```diff
@@ -257,4 +257,308 @@ TEST_P(metics_c_p, test_metric_04) {
+INSTANTIATE_TEST_CASE_P(metrics_c, metrics_c_p, ::testing::ValuesIn(test_platform::mock_platforms({ "skx-p" })));
+
+
+int mmio_ioctl(mock_object * m, int request, va_list argp) {
+	int retval = -1;
+	errno = EINVAL;
+	UNUSED_PARAM(m);
+	UNUSED_PARAM(request);
+	struct fpga_port_region_info *rinfo = va_arg(argp, struct fpga_port_region_info *);
+	if (!rinfo) {
+		FPGA_MSG("rinfo is NULL");
+		goto out_EINVAL;
+	}
+	if (rinfo->argsz != sizeof(*rinfo)) {
+		FPGA_MSG("wrong structure size");
+		goto out_EINVAL;
+	}
+	if (rinfo->index > 1) {
+		FPGA_MSG("unsupported MMIO index");
+		goto out_EINVAL;
+	}
+	if (rinfo->padding != 0) {
+		FPGA_MSG("unsupported padding");
+		goto out_EINVAL;
+	}
+	rinfo->flags = FPGA_REGION_READ | FPGA_REGION_WRITE | FPGA_REGION_MMAP;
+	rinfo->size = 0x40000;
+	rinfo->offset = 0;
+	retval = 0;
+	errno = 0;
+
+out:
+	return retval;
+
+out_EINVAL:
+	retval = -1;
+	errno = EINVAL;
+	goto out;
+}
+
+
+
+
+/**
+* @brief metrics afu gtest fixture
+*
+*/
+class metrics_afu_c_p : public ::testing::TestWithParam<std::string> {
+protected:
+	metrics_afu_c_p()
+		: tokens_{ {nullptr, nullptr} },
+		handle_(nullptr) {}
+	void create_metric_bbb_dfh();
+	void create_metric_bbb_csr();
+	virtual void SetUp() override {
+
+		ASSERT_TRUE(test_platform::exists(GetParam()));
+		platform_ = test_platform::get(GetParam());
+		system_ = test_system::instance();
+		system_->initialize();
+		system_->prepare_syfs(platform_);
+
+		ASSERT_EQ(xfpga_fpgaGetProperties(nullptr, &filter_), FPGA_OK);
+		ASSERT_EQ(fpgaPropertiesSetObjectType(filter_, FPGA_ACCELERATOR), FPGA_OK);
+		ASSERT_EQ(xfpga_fpgaEnumerate(&filter_, 1, tokens_.data(), tokens_.size(),
+			&num_matches_),
+			FPGA_OK);
+
+		ASSERT_EQ(xfpga_fpgaOpen(tokens_[0], &handle_, 0), FPGA_OK);
+
+
+		system_->register_ioctl_handler(FPGA_PORT_GET_REGION_INFO, mmio_ioctl);
+		which_mmio_ = 0;
+		uint64_t *mmio_ptr = nullptr;
+		EXPECT_EQ(xfpga_fpgaMapMMIO(handle_, which_mmio_, &mmio_ptr), FPGA_OK);
+		EXPECT_NE(mmio_ptr, nullptr);
+	}
+
+	virtual void TearDown() override {
+		EXPECT_EQ(xfpga_fpgaUnmapMMIO(handle_, which_mmio_), FPGA_OK);
+		EXPECT_EQ(fpgaDestroyProperties(&filter_), FPGA_OK);
+		for (auto &t : tokens_) {
+			if (t) {
+				EXPECT_EQ(xfpga_fpgaDestroyToken(&t), FPGA_OK);
+				t = nullptr;
+			}
+		}
+		if (handle_) { EXPECT_EQ(xfpga_fpgaClose(handle_), FPGA_OK); }
+		system_->finalize();
+	}
+	uint32_t which_mmio_;
+	std::array<fpga_token, 2> tokens_;
+	fpga_handle handle_;
+	fpga_properties filter_;
+	uint32_t num_matches_;
+	test_platform platform_;
+	test_system *system_;
+
+};
+
+void metrics_afu_c_p::create_metric_bbb_dfh() {
+
+	struct DFH dfh;
+	dfh.id = 0x1;
+	dfh.revision = 0;
+	dfh.next_header_offset = 0x100;
+	dfh.eol = 1;
+	dfh.reserved = 0;
+	dfh.type = 0x1;
+
+
+	printf("------dfh.csr = %lx \n", dfh.csr);
+	EXPECT_EQ(FPGA_OK, xfpga_fpgaWriteMMIO64(handle_, 0, 0x0, dfh.csr));
+
+	EXPECT_EQ(FPGA_OK, xfpga_fpgaWriteMMIO64(handle_, 0, 0x8, 0xf89e433683f9040b));
```

Comment:
Can this be made agnostic to the type of AFU somehow using the test framework?

Reply:
this codes adds AFU  Header & Metric BBB in MMIO region.